### PR TITLE
Add support for TypeScript

### LINKED ISSUE #113: Add support for TypeScript
[Probot](https://github.com/probot/probot/tree/master/src/%40types/%40octokit/plugin-throttling) already includes the definitions. We could just include them here and publish them with the package.

### PR DESCRIPTION
**Pull request by douglascayers**
Fixes #113

### CODE REVIEW

File: @types/@octokit/plugin-throttling/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+declare module '@octokit/plugin-throttling' {
+  import Octokit from '@octokit/rest'
+  function plugin(octokit: Octokit, options: Octokit.Options): void
+  export = plugin
+}
```
Comment (gr2m) [inline @types/@octokit/plugin-throttling/index.d.ts]: I wonder why you don't put an `index.d.ts` into the root folder and set the `"types"` key in package.json? Similar to what we do here https://github.com/octokit/rest.js/blob/9b85cd674741ee67d03f510e19d24a9d14014e02/package.json#L81

Comment (douglascayers) [reply @types/@octokit/plugin-throttling/index.d.ts]: Hi @gr2m, this is based solely on the link that the issue originally referenced: https://github.com/probot/probot/tree/master/src/%40types/%40octokit/plugin-throttling

Assuming that’s the type, I just opened a PR for it. 

Disclaimer, I’m not a typescript expert and muddling my way through it. Please correct as appropriate.

Thanks!

Comment (douglascayers) [reply @types/@octokit/plugin-throttling/index.d.ts]: @gr2m, I did some more review and realized the `index.d.ts` I proposed didn't actually provide any interfaces or type definitions at all. I've pushed new commits as well as moved the file to the root folder per your suggestion. Thanks

Comment (gr2m) [reply @types/@octokit/plugin-throttling/index.d.ts]: > I’m not a typescript expert and muddling my way through it

Same 😄

**Comment by gr2m** [comment]
Did you try if it works? Do you get a typeahead for `options.throttle` when applying the plugin?

For `@octokit/rest` we added a typescript test, a single file that we evaluate with `tsc` to make sure there are no syntax errors, as a means to prevent accidental changes in the type definitions. Maybe you could add one to this repo, too?

https://github.com/octokit/rest.js/blob/master/test/typescript-validate.ts

**Comment by douglascayers** [comment]
I get your checking in my project where I copied this file. I’ll explore adding a unit test.

**Comment by leafac** [comment]
@douglascayers Thank you for picking up where I left off 😃

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+declare module '@octokit/plugin-throttling' {
+    import Octokit from '@octokit/rest';
+    import Bottleneck from 'bottleneck';
+
+    /**
+     * More context provided to your limit callback function.
+     */
+    interface LimitCallbackParams {
+        request: {
```
Comment (Fl0pZz) [inline index.d.ts]: Example from README:
```js
onAbuseLimit: (retryAfter, options) => {
    // does not retry, only logs a warning
    console.warn(`Abuse detected for request ${options.method} ${options.url}`)
}
```
`request` may have more properties

**Comment by gr2m** [comment]
superseded by #125. But maybe some of the types from Doug's PR could be migrated? The comments are great

**Comment by gr2m** [comment]
Closing the issue to better keep track of pending work

**Pull request closed without merging**